scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -279,9 +279,121 @@
   (-> (d/entity db [:instance/task-id task-id])
       :db/id))
 
+(histograms/defhistogram [cook-mesos scheduler progress-aggregator-pending-states])
+(meters/defmeter [cook-mesos scheduler progress-aggregator-message-rate])
+(meters/defmeter [cook-mesos scheduler progress-aggregator-drop-rate])
+(counters/defcounter [cook-mesos scheduler progress-aggregator-drop-count])
+
+(defn progress-update-aggregator
+  "Launches a long running go block that triggers publishing the latest aggregated instance-id->progress-state
+   whenever there is a message on the progress-aggregator-trigger-chan.
+   It drops messages if the progress-aggregator-chan queue is larger than pending-progress-threshold or the aggregated
+   state has more than pending-progress-threshold different entries.
+   It returns a map containing progress-aggregator-chan and query-state-chan.
+   progress-aggregator-chan can be used to send progress-state messages to the aggregator.
+   query-state-chan can be used to query the state of the aggregator."
+  [pending-progress-threshold progress-aggregator-trigger-chan progress-state-chan]
+  (log/info "Starting progress update aggregator")
+  (let [progress-aggregator-chan (async/chan (async/sliding-buffer pending-progress-threshold))
+        query-state-chan (async/chan (async/sliding-buffer 1))]
+    (async/go
+      (loop [instance-id->progress-state {}]
+        (let [[data chan] (async/alts! [progress-aggregator-trigger-chan progress-aggregator-chan query-state-chan]
+                                       :priority true)]
+          (condp = chan
+            progress-aggregator-trigger-chan
+            (do
+              (when (seq instance-id->progress-state)
+                (log/debug "Forwarding" (count instance-id->progress-state) "progress states to the transactor")
+                (histograms/update! progress-aggregator-pending-states (count instance-id->progress-state))
+                (async/>! progress-state-chan instance-id->progress-state))
+              (recur {}))
+
+            progress-aggregator-chan
+            (let [{:keys [instance-id]} data]
+              (meters/mark! progress-aggregator-message-rate)
+              (if (or (< (count instance-id->progress-state) pending-progress-threshold)
+                      (contains? instance-id->progress-state instance-id))
+                (let [progress-state (select-keys data [:progress-message :progress-percent])
+                      instance-id->progress-state' (assoc instance-id->progress-state instance-id progress-state)]
+                  (recur instance-id->progress-state'))
+                (do
+                  (meters/mark! progress-aggregator-drop-rate)
+                  (counters/inc! progress-aggregator-drop-count)
+                  (log/debug "Dropping" data "as threshold has been reached")
+                  (recur instance-id->progress-state))))
+
+            query-state-chan
```

Comment:
I don't see where this is used except in tests. This seems like unneeded complexity. If you remove the trigger channel in this loop and allow for pulling you will be able to get the query, though it will wipe the state and then you could close the progress-aggregator-chan to signal exit.